Sources/SwiftFusion/Inference/Metropolis.swift

```diff
@@ -0,0 +1,83 @@
+// Copyright 2020 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import PenguinGraphs
+
+/// Metropolis algorithm.
+extension FactorGraph where
+  // The Metropolis algorithm needs to query the graph for neighboring vertices:
+  Graph: IncidenceGraph,
```

Comment:
Here you actually mean incidence to variables, right? But we might also need incidence from factors to variables. I think agree with the statement from the call that a bipartite graph will almost never be used  as a “regular” graph. But one thing you can do is to calculate a regular graph on the variables *from* the factor graph.

BTW, there is a simple way to explain this in terms of sparse matrices: if a sparse matrix A encodes the non-zero pattern of a factor graph, then A^T A is the undirected (symmetric!) adjacency graph between variables, and AA^T is the adjacency graph between factors.

So, your metropolis proposal (really the only thing that wants adjacent variables) could use a *computed* adjacency graph which is a regular undirected graph, not a bipartite graph.

Reply:
The metropolis proposal might want variable adjacency information, yes. (Or it might not, depending on what the proposal is.)

However, this implementation of metropolis is completely agnostic to the proposal. This implementation just takes the proposal as a closure.

This implementation of metropolis uses the graph structure to determine which factors must be recomputed to calculate the acceptance ratio. Specifically, it looks at the proposal, sees which variables have changed, and only recomputes factors adjacent to those variables.

This implementation is only one point in the design space of generic metropolis algorithms. It could be a totally wrong point in the design space. Perhaps the proposal should be responsible for calculating the set of affected factors, so that the generic metropolis algorithm doesn't need to inspect the graph structure itself.